Initial docs (#994)
redhat-developer/rhdh-plugin-export-overlays: README.md

## What is the rhdh-plugin-export-overlays reposiotry?

> The rhdh-plugin-export-overlays repository serves as a metadata and automation hub for managing dynamic plugins in Red Hat Developer Hub (RHDH).

Reviewer: ```suggestion
The rhdh-plugin-export-overlays repository serves as a metadata and automation hub for managing dynamic plugins for Red Hat Developer Hub (RHDH).
```

This repository:

- References a wide range of Backstage plugins that can or should be published as dynamic plugins for use in RHDH.

- Tracks plugin versions to ensure compatibility with the latest RHDH releases.

- Defines how to drive, customize, and automate the publishing process.

Additionally, it contains workflows to:

- Discover eligible Backstage plugins.

- Package them as OCI images for use as dynamic plugins.

- Publish these images to the GitHub Container Registry for easy integration with RHDH.

## Repository Structure & Partitioning
The content in this repository is structured as follows:

1. **By Workspace:**

> Each plugin set is organized in a dedicated folder that represents a workspace—typically aligned with a monorepo hosted in a third-party GitHub repository (e.g., @backstage-community, @roadiehq, @red-hat-developer-hub).

Reviewer: ```suggestion
Each plugin set is organized in a dedicated folder that represents a workspace—typically aligned with a monorepo hosted in a third-party GitHub repository (e.g., `@backstage-community`, `@roadiehq`, `@red-hat-developer-hub`).
```

1. **By RHDH Target Release:**

> The repository uses long-running release branches named following the pattern release-x.y (e.g., release-1.5, release-1.6).

Reviewer: ```suggestion
The repository uses long-running release branches named following the pattern `release-x.y` (e.g., `release-1.5`, `release-1.6`).
```

   These PRs are created **per workspace, per release branch**.
   
   If you can't find a PR for your plugin, you can manually trigger one as explained below.
   
   ### Create a PR using the "Update plugins repository references" workflow

  > [!IMPORTANT]
  > Write access to this repository is required to run this workflow.

   - Navigate to https://github.com/redhat-developer/rhdh-plugin-export-overlays/actions/workflows/update-plugins-repo-refs.yaml 
   - For "use workflow from" select main
   - For "regexps", specify the regular expression matching the plugins you want to package. For example, to package all RBAC plugins, the regexp would be "@backstage-community/plugin-rbac"
   - Running the workflow with the following inputs should generate a PR similar to [these ones](https://github.com/redhat-developer/rhdh-plugin-export-overlays/pulls/app%2Fgithub-actions)
    
   **How it works**:
   - The workflow generates PRs against each active release branch corresponding to an RHDH release.

>    - For each release branch, it checks for plugin versions compatible with the Backstage version that the release supports.

Reviewer: ```suggestion
   - For each release branch, it checks for published plugin versions compatible with the Backstage version that the RHDH release supports.
```

3. Add a `source.json` file with the following fields ([See example](https://github.com/redhat-developer/rhdh-plugin-export-overlays/blob/12acb71a1febc5567c4d12c6a28c0a11ed489273/workspaces/adoption-insights/source.json))
:
   - `repo`: URL of the source repository  (only `https://github.com/xxx` URLs are supported for now)
   - `repo-ref`: Specific tag or commit for the target plugin/workspace version  
   - `repo-flat`:  
     - `false` if the plugins are inside a workspace (e.g., `backstage/community-plugins`)  
     - `true` if the plugins are at the root level (e.g., `backstage/backstage`)

### 3. Add Additional Dynamic Plugin Export Information (If Needed)

Sometimes, additional configuration is required in the PR:

- **Frontend plugins** may need:
   - `app-config.dynamic.yaml` (Eg: [techdocs plugin](https://github.com/redhat-developer/rhdh-plugin-export-overlays/blob/release-1.5/workspaces/backstage/plugins/techdocs/app-config.dynamic.yaml))

>    - `scalprum-config.json` (Eg: [api-docs-module-protoc-gen-doc plugin](https://github.com/redhat-developer/rhdh-plugin-export-overlays/blob/release-1.5/workspaces/backstage/plugins/api-docs-module-protoc-gen-doc/scalprum-config.json))

Reviewer: In most case this one should not be necessary, especially due to the fact that, in the future, we would move away from Scalprum.
Not sure we should even mention this here.

Reviewer: Or maybe just mention that this is only rarely required ?

- **Any plugin** may need:
   - Overlay source files in an `overlay` directory  
  (e.g., [`api-docs-module-protoc-gen-doc`](https://github.com/redhat-developer/rhdh-plugin-export-overlays/tree/release-1.5/workspaces/backstage/plugins/api-docs-module-protoc-gen-doc/overlay))

>   - A patch file at the root of the workspace to modify plugin source code during the packaging process. (Example: PR [#792](https://github.com/redhat-developer/rhdh-plugin-export-overlays/pull/792/files#diff-0b648cbca6f87e11f78832c10ac9cc789235938e944c499eb275fd8788e18ef8))

Reviewer: ```suggestion
  - A patch file at the root of the workspace to modify the workspace source code before the packaging process. (Example: PR [#792](https://github.com/redhat-developer/rhdh-plugin-export-overlays/pull/792/files#diff-0b648cbca6f87e11f78832c10ac9cc789235938e944c499eb275fd8788e18ef8))
```

> **Overlay vs. Patch**  
> - **Overlay**: Replaces or adds entire files during the packaging process.  
> - **Patch**: Applies precise, line-by-line changes to existing source files.  

> To add this:

Reviewer: When adding a patch, you don't need to create the plugins/<plugin-name> folder since the patch file is added at the workspace (== monorepo) root.

> To add this additional configration (excluding the patch, since the patch file is placed at the workspace [monorepo] root):

Reviewer: ```suggestion
To add this additional configuration (excluding the patch, since the patch file is placed at the workspace [monorepo] root):
```

## What is the rhdh-plugin-export-overlays reposiotry?

This repository contains workflows to package Backstage plugins (under the following npm scopes: @backstage-community, @red-hat-developer-hub and @roadiehq) as OCI images to be used as dynamic plugins within Red Hat Developer Hub (RHDH).

> 

Reviewer: Maybe some additional explanation about the fact that the content of the data is partitioned:
- per workspace, in distinct folders corresponding to disctinct monorepos hosted in some third-party GH repositories)
- per RHDH target release, based on the use of release branches: long-running branches named `release-x.y`, x.y being the RHDH release (1.5, 1.6, etc ).

## How to use the workflows in this repository to create OCI images for your plugins

### 1. Create/Look for a PR for the plugins you wish to export against the desired RHDH relase branch version 
   

>    A GitHub workflow runs **daily at 12:00 PM UTC** to automatically generate or update PRs for plugins under the supported scopes. These PRs are created **per workspace, per release branch**.

Reviewer: ```suggestion
   A GitHub workflow runs **daily at 12:00 PM UTC** to automatically generate or update PRs for plugins under the following scope: `@backstage-community`, `@red-hat-developer-hub` and `@roadiehq`.
   These PRs are created **per workspace, per release branch**.
```
